dev/NumberBox/NumberBoxParser.cpp

```diff
@@ -0,0 +1,301 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+#include "common.h"
+#include "NumberBoxParser.h"
+#include "Utils.h"
+
+static constexpr wstring_view c_numberBoxOperators{ L"+-*/^"sv };
+
+// Returns list of MathTokens from expression input string. If there are any parsing errors, it returns an empty vector.
+std::vector<MathToken> NumberBoxParser::GetTokens(const wchar_t* input, winrt::INumberParser numberParser)
+{
+    auto tokens = std::vector<MathToken>();
+
+    size_t index = 0;
+    bool error = false;
+    bool expectNumber = true;
+    while (input[0] != '\0' && !error)
+    {
+        // Skip spaces
+        auto nextChar = input[0];
+        if (nextChar != L' ')
+        {
+            if (expectNumber)
+            {
+                if (nextChar == '(')
+                {
+                    // Open parens are also acceptable, but don't change the next expected token type.
+                    tokens.push_back(MathToken(MathTokenType::Parenthesis, nextChar));
+                }
+                else
+                {
+                    double value = NAN;
+                    const auto charLength = GetNextNumber(input, numberParser, value);
+
+                    if (charLength > 0)
+                    {
+                        tokens.push_back(MathToken(MathTokenType::Numeric, value));
+                        input += charLength - 1; // advance the end of the token
+                        expectNumber = false; // next token should be an operator
+                    }
+                    else
+                    {
+                        // Error case -- next token is not a number
+                        error = true;
+                    }
+                }
+            }
+            else
+            {
+                if (c_numberBoxOperators.find(nextChar) != std::wstring::npos)
+                {
+                    tokens.push_back(MathToken(MathTokenType::Operator, nextChar));
+                    expectNumber = true; // next token should be a number
+                }
+                else if (nextChar == ')')
+                {
+                    // Closed parens are also acceptable, but don't change the next expected token type.
+                    tokens.push_back(MathToken(MathTokenType::Parenthesis, nextChar));
+                }
+                else
+                {
+                    // Error case -- could not evaluate part of the expression
+                    error = true;
+                }
+            }
+        }
+
+        input++;
+    }
+
+    if (error)
+    {
+        return {};
+    }
+    return tokens;
+}
+
+// Attempts to parse a number from the beginning of the given input string. Returns the character size of the matched string.
+size_t NumberBoxParser::GetNextNumber(std::wstring input, winrt::INumberParser numberParser, double& outValue)
```

Comment:
>double& outValue [](start = 93, length = 16)

I think out parameters should only be used if you need to manipulate the state of an object being handed in.  In this case you are completely overriding what may have been passed into the function as outValue and thus I think it would be better for this function to return a tuple which contains the size and the value.

Reply:
I didn't know that was a thing! Cool, I like that.

Reply:
I couldn't find the CPP guideline that says this until now, it is F.21

---
In reply to: [345446862](https://github.com/microsoft/microsoft-ui-xaml/pull/1602#discussion_r345446862) [](ancestors = 345446862)